vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,318 @@
-                })
-            })
-            .map_err(|e| {
+        if let Some(err) = act_err {
+            if err.is_fatal() {
+                return Err(format!("Fatal send actor error occurred, err: {:?}", err));
+            };
+            if err.exit_code() != ExitCode::Ok {
+                // revert all state changes since snapshot
                 if let Err(state_err) = self.state.revert_to_snapshot(&snapshot) {
-                    return state_err;
-                }
-                e.to_string()
-            })?
+                    return Err(format!("Revert state failed: {}", state_err));
+                };
+            }
+            warn!("Send actor error: from:{}, to:{}", msg.from(), msg.to());
+        }
+        // TODO ask if this is correct as this could give a free tx if gas used was a negative value(e.g. delete_miner)
+        // see: https://github.com/filecoin-project/lotus/blob/testnet/3/chain/vm/vm.go#L388
+        if gas_used < 0 {
+            gas_used = 0;
+        }
+        // refund unused gas
+        let refund =
+            (msg.gas_limit().checked_sub(gas_used as u64).unwrap_or(1u64)) * msg.gas_price();
+        transfer_from_gas_holder(&mut self.state, msg.from(), &mut gas_holder, &refund)?;
+
+        let gas_reward = msg.gas_price() * BigUint::from(gas_used as u64);
+        transfer_from_gas_holder(
+            &mut self.state,
+            &*REWARD_ACTOR_ADDR,
+            &mut gas_holder,
+            &gas_reward,
+        )?;
+
+        if gas_holder.balance != BigUint::zero() {
+            return Err("Gas handling math is wrong".to_owned());
+        }
+
+        Ok(ApplyRet::new(
+            MessageReceipt {
+                return_data: ret_data,
+                exit_code: ExitCode::Ok,
+                gas_used: BigUint::from(gas_used as u64),
+            },
+            BigUint::zero(),
+            None,
+        ))
     }
     /// Instantiates a new Runtime, and calls internal_send to do the execution.
-    fn send(&mut self, msg: &UnsignedMessage, gas_cost: i64) -> Result<Serialized, ActorError> {
+    fn send<'m>(
+        &mut self,
+        msg: &'m UnsignedMessage,
+        gas_cost: i64,
+    ) -> (
+        Serialized,
+        DefaultRuntime<'_, 'm, '_, ST, DB, SYS>,
+        Option<ActorError>,
+    ) {
         let mut rt = DefaultRuntime::new(
             &mut self.state,
             self.store,
-            DefaultSyscalls,
+            self.syscalls,
             gas_cost,
             &msg,
             self.epoch,
             msg.from().clone(),
             msg.sequence(),
             0,
         );
-        internal_send(&mut rt, msg, gas_cost)
+
+        let ser = match internal_send(&mut rt, msg, gas_cost) {
+            Ok(ser) => ser,
+            Err(actor_err) => return (Serialized::default(), rt, Some(actor_err)),
+        };
+        (ser, rt, None)
+    }
+}
+
+// TODO remove allow dead_code
+#[allow(dead_code)]
```

Comment:
Remove and add a doc comment? or is still dead code because fields are not read?

Reply:
Still, fields are not read.